langserver/config.go

```diff
@@ -137,41 +154,91 @@ func readConfigFromENV() (*Config, error) {
+	if params == nil {
+		return nil
+	}
+	// nolint: errcheck
+	s.client.LogMessage(
+		s.lifetime,
+		&protocol.LogMessageParams{
+			Type:    protocol.Info,
+			Message: fmt.Sprintf("Received notification change: %v\n", params),
+		})
+
+	setting := params.Settings
+
+	// the struct expected is the following
+	// promql:
+	//   url: http://
+	//   interval: 3w
+	m, ok := setting.(map[string]map[string]string)
+	if !ok {
+		// nolint: errcheck
+		s.client.LogMessage(ctx, &protocol.LogMessageParams{
+			Type:    protocol.Error,
+			Message: fmt.Sprint("unexpected format of the configuration"),
+		})
+		return nil
+	}
+	config, ok := m["promQL"]
+	if !ok {
+		// nolint: errcheck
+		s.client.LogMessage(ctx, &protocol.LogMessageParams{
+			Type:    protocol.Error,
+			Message: fmt.Sprint("promQL key not found"),
+		})
+		return nil
+	}
 
-	if params != nil {
+	if err := s.setURLFromChangeConfiguration(config); err != nil {
 		// nolint: errcheck
-		s.client.LogMessage(
-			s.lifetime,
-			&protocol.LogMessageParams{
-				Type:    protocol.Info,
-				Message: fmt.Sprintf("Received notification change: %v\n", params),
-			})
-
-		setting := params.Settings
-
-		for _, e := range langserverAddressConfigPath {
-			m, ok := setting.(map[string]interface{})
-			if !ok {
-				break
-			}
-
-			setting, ok = m[e]
-			if !ok {
-				break
-			}
-		}
+		s.client.LogMessage(ctx, &protocol.LogMessageParams{
+			Type:    protocol.Info,
+			Message: err.Error(),
+		})
+	}
 
-		if str, ok := setting.(string); ok {
-			if err := s.connectPrometheus(str); err != nil {
-				// nolint: errcheck
-				s.client.LogMessage(ctx, &protocol.LogMessageParams{
-					Type:    protocol.Info,
-					Message: err.Error(),
-				})
-			}
+	if err := s.setMetadataLookbackIntervalFromChangeConfiguration(config); err != nil {
+		// nolint: errcheck
+		s.client.LogMessage(ctx, &protocol.LogMessageParams{
+			Type:    protocol.Info,
+			Message: err.Error(),
+		})
+	}
+	return nil
+}
+
+func (s *server) setURLFromChangeConfiguration(settings map[string]string) error {
+	if promURL, ok := settings["url"]; ok {
+		if _, err := url.Parse(promURL); err != nil {
+			return err
+		}
+		if err := s.connectPrometheus(promURL); err != nil {
+			return err
 		}
 	}
+	return nil
+}
+
+func (s *server) connectPrometheus(url string) error {
+	if err := s.metadataService.ChangeDataSource(url); err != nil {
+		// nolint: errcheck
+		s.client.ShowMessage(s.lifetime, &protocol.ShowMessageParams{
+			Type:    protocol.Error,
+			Message: fmt.Sprintf("Failed to connect to Prometheus at %s:\n\n%s ", url, err.Error()),
+		})
+		return err
+	}
+	return nil
+}
 
+func (s *server) setMetadataLookbackIntervalFromChangeConfiguration(settings map[string]string) error {
+	if interval, ok := settings["metadataLookbackInterval"]; ok {
+		duration, err := model.ParseDuration(interval)
+		if err != nil {
+			return err
+		}
+		s.config.MetadataLookbackInterval = duration
```

Comment:
This looks like a race condition.

Reply:
yeah I did it on purpose. It doesn't require to have mutex here because at some point it will become stable. So it's not really important that some queries are using an old interval.

Reply:
There's so much more that can go wrong.

For an introduction, this article (about C, not Go) is a good start.

https://lwn.net/Articles/793253/

Reply:
> I'm not really in favor of doing the same than the prometheus URL. It doesn't require the same completexity and I prefer that the `metadataService` is opened to use freely the from/start value instead of passing by an attribute.

Looking up the interval in a thread safe way, will complexity and duplication that shouldn't be in the completion/hover code.

Details like lookback time, which may not even make sense for certain future data sources, should be put into the `metadataService` abstraction.

Reply:
>future datas ources 

Like which one ?

Excepting if you are planning to support promQL from something else than prometheus, I don't see what you mean

>For an introduction, this article (about C, not Go) is a good start.
Like you said, that's C not Go. Compilation is not the same :/. Even in Java you could do that if you are ok to have an unstable state at a certain point which will be stable after some times

Reply:
And actually I think I should put a pointer instead  of a plain value. Like that the adress to the value won't change and it will be a bit more stable.

Reply:
Go in some cases uses a gcc (cgo) or llvm (tinygo) backend, which gives us all the c compiler fun.

Anyway, I'm strictly against intentionally introducing race conditions out of laziness, so there is no way this PR will get merged until this is fixed.

Reply:
> And actually I think I should put a pointer instead  of a plain value. Like that the adress to the value won't change and it will be a bit more stable.

And risk load/store tearing pointers? That's worse.

Reply:
To make clearer, why the risk of race conditions in the language server is so high, this is how RPC requests are handled:

https://github.com/prometheus-community/promql-langserver/blob/10b1d7e0b5f6fd3ed6fe53776aedf56b17a889db/internal/vendored/go-tools/jsonrpc2/jsonrpc2.go#L364-L384

Reply:
mmm 
1. So first I think I will forget you were implying I am lazy.
2. Then I will rework a bit the way the configuration is handled because configuration should be only readable. It's just matter of struct usage and so on. So it won't change the whole thing, it will just enforce the responsability of each layer/object.

Don't know If I will do it in this PR or in another. Maybe in another will be easier to read/understand

3. After that then it should be easier to make the access to this attribute thread safe.

Reply:
and then let's see if it makes more sense to have this `interval` on the prometheus client side or in the server side. Just matter of how it will be easier to use on Prometheus when integrated

Reply:
>     2. Then I will rework a bit the way the configuration is handled because configuration should be only readable.

Changing configuration at runtime (without restarting) is a feature of the language server protocol and of VS Code.

Reply:
> and then let's see if it makes more sense to have this `interval` on the prometheus client side or in the server side. Just matter of how it will be easier to use on Prometheus when integrated

Maybe the people reviewing that PR on the Prometheus side should be involved in that discussion.

Reply:
> Changing configuration at runtime (without restarting) is a feature of the language server protocol and of VS Code.

Yes I know. But that's another configuration. So the "cold" configuration should be treated like that. I wouldn't say that if it was the same config.

Basically I'm just saying we should move outside of the package the yaml config. This config will be used to fill some attribute on the server. Same process will be apply to the "hot" config coming from the method `DidChangeConfiguration`

Two different conf, two different struct, two different way to treat it. But at the end it will fill the same attributes of the server.

Reply:
That approach sounds good to me.

Reply:
and the struct that is handling the configuration is usually used to just fill some attribute of a particular other struct. But we generally don't use it directly in an object because well you don't need everything in the config. So you just take what you need and then you drop it.

But you will see It should make more sense once the PR is ready ;)

Reply:
> Maybe the people reviewing that PR on the Prometheus side should be involved in that discussion.

Perhaps but I'm feeling it will come naturally when coding it. So let's see if we need to involve more people here